### ISSUE TITLE
Crash when entering single acute into IMP website password

### ISSUE THREAD

**Issue by Manuel-K** [ja2-stracciatella/ja2-stracciatella#497]
The game segfaults after printing the messages
```
terminate called after throwing an instance of 'InvalidEncodingException'
  what():  Invalid UTF-8 encoding
```

This happens on Arch Linux with XkbLayout = de and XkbVariant = nodeadkeys.

**Comment by JAsmine-ja2** [comment]
In vanilla, we can't enter a single acute (in fact, we can't enter anything special besides underlines and periods), so the bug can't be reproduced there.

I also can't reproduce the crash with Stracciatella 0.15.1 on Windows. I can enter single acutes without crashing.

**Comment by Manuel-K** [comment]
The problem is more general than I thought.  If one of the problematic keys (not only ´, also ü, ö, ä and many other keys) is pressed, the game crashes anywhere. Even in the main menu. At first I suspected `nodeadkeys` to be the source of the problem, but that was clearly wrong.


Note: I'm using master, i.e. db7fa39972e61cbb58a39e6e210c23220cdb7c0b.

**Comment by Manuel-K** [comment]
I'm still not sure why it does crash on my system and not on @JAsmine-ja2's, but the problem is in `void TextInput(const SDL_TextInputEvent* TextEv)` in `src/sgp/Input.cc`.

**Comment by Manuel-K** [comment]
Should we add an `try{}`, `catch(){}` around the functions body? That would ignore invalid inputs.

**Comment by JAsmine-ja2** [comment]
I can't help with a solution, but, for the record, this is how it looks here:

![lap1](https://cloud.githubusercontent.com/assets/17213795/23338391/00b4dc48-fc0a-11e6-8c57-56a1e8bd1695.jpg)
![lap2](https://cloud.githubusercontent.com/assets/17213795/23338392/00d7f61a-fc0a-11e6-8f5a-a43d440e5941.jpg)

**Comment by Manuel-K** [comment]
That's very strange.

The problem is either in `src/sgp/UTF8String.cc` or it is in the generation of the `SDL_TextInputEvent`. Either way, adding a try, catch ignores the problem and I think that is a good idea, because pressing a wrong key should not crash the game on any platform.

**Comment by selaux** [comment]
Yes, but entering `ö` should enter `ö` into the input 🙂 . You shouldn't be able to enter a character that is not UTF16-compatible. I have almost the same configuration and pressing these buttons does not crash for me. What is the output of `locale` for you?

**Comment by Manuel-K** [comment]
It's what I would expect:
```
LANG=de_DE.UTF_8
LC_CTYPE="de_DE.UTF_8"
LC_NUMERIC="de_DE.UTF_8"
LC_TIME="de_DE.UTF_8"
LC_COLLATE=C
LC_MONETARY="de_DE.UTF_8"
LC_MESSAGES="de_DE.UTF_8"
LC_PAPER="de_DE.UTF_8"
LC_NAME="de_DE.UTF_8"
LC_ADDRESS="de_DE.UTF_8"
LC_TELEPHONE="de_DE.UTF_8"
LC_MEASUREMENT="de_DE.UTF_8"
LC_IDENTIFICATION="de_DE.UTF_8"
LC_ALL=
```

**Comment by Manuel-K** [comment]
What SDL version do you use? I'm using version 2.0.5.

**Comment by selaux** [comment]
Same here. Can you create a breakpoint inside `TextInput` in `Input.cc` to see what happens if you enter one of these characters?

**Comment by Manuel-K** [comment]
After pressing k:
```
	Locals		
		TextEv	@0x7fffffffe4b8	SDL_TextInputEvent
			text	"k\000\000\000\000\021\000A\000\000\000\000\000���\020����\013\000A\000\000\000\0004\014\000A"	char[32]
					'k' 	107    	0x6b	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					    	17    	0x11	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					    	-1/255	0xff	char
					'￨' 	-24/232	0xe8	char
					'ￎ' 	-50/206	0xce	char
					    	16    	0x10	char
					    	-55/201	0xc9	char
					    	-14/242	0xf2	char
					'ﾼ' 	-68/188	0xbc	char
					'ﾀ' 	-128/128	0x80	char
					    	11    	0x0b	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'4' 	52    	0x34	char
					    	12    	0x0c	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
			timestamp	3183	Uint32
			type	771	Uint32
			windowID	2	Uint32
		utf8String	@0x7fffffffe4c0	UTF8String
	Inspector		
	Expressions		
	Return Value		
	Tooltip		
```

After pressing ö:
```
	Locals		
		TextEv	@0x7fffffffe4b8	SDL_TextInputEvent
			text	"�\000\000\000\000\021\000A\000\000\000\000\000\003�\025�����\013\000A\000\000\000\000<\014\000A"	char[32]
					    	-10/246	0xf6	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					    	17    	0x11	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					    	3    	0x03	char
					'ﾔ' 	-108/148	0x94	char
					    	21    	0x15	char
					'ﾎ' 	-114/142	0x8e	char
					'ﾒ' 	-110/146	0x92	char
					    	-6/250	0xfa	char
					    	-65/191	0xbf	char
					'ﾀ' 	-128/128	0x80	char
					    	11    	0x0b	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'\0'	0    	0x00	char
					'<' 	60    	0x3c	char
					    	12    	0x0c	char
					'\0'	0    	0x00	char
					'A' 	65    	0x41	char
			timestamp	1615	Uint32
			type	771	Uint32
			windowID	2	Uint32
		utf8String	@0x7fffffffe4c0	UTF8String
	Inspector		
	Expressions		
	Return Value		
	Tooltip		
```

0xf6 seems to be right.

**Comment by selaux** [comment]
So the resulting utf8String is a single `ö`?

**Comment by Manuel-K** [comment]
No, it's "". The exception prevents the assignment.

**Comment by selaux** [comment]
You can try showing where encoding fails if you call the method `utf8_octet` of the resulting error...

**Comment by Manuel-K** [comment]
I get an `InvalidEncodingException`, not an `invalid_utf8` which has the method `utf8_octet`.



A breakpoint on this line gets hit:
https://github.com/ja2-stracciatella/ja2-stracciatella/blob/master/dependencies/lib-utf8cpp/source/utf8/core.h#L141


```
1  utf8::internal::increase_safely<__gnu_cxx::__normal_iterator<char *, std::string>> core.h        141 0x82c104 
2  utf8::internal::get_sequence_4<__gnu_cxx::__normal_iterator<char *, std::string>>  core.h        205 0x82be3b 
3  utf8::internal::validate_next<__gnu_cxx::__normal_iterator<char *, std::string>>   core.h        249 0x82b3ab 
4  utf8::internal::validate_next<__gnu_cxx::__normal_iterator<char *, std::string>>   core.h        277 0x82b153 
5  utf8::find_invalid<__gnu_cxx::__normal_iterator<char *, std::string>>              core.h        292 0x82a409 
6  utf8::is_valid<__gnu_cxx::__normal_iterator<char *, std::string>>                  core.h        302 0x829dfe 
7  UTF8String::UTF8String                                                             UTF8String.cc 11  0x829281 
8  TextInput                                                                          Input.cc      359 0x8189aa 
9  MainLoop                                                                           SGP.cc        217 0x81fbcb 
10 main                                                                               SGP.cc        474 0x820718 
```

**Comment by Manuel-K** [comment]
Okay. I'm a bit closer to the solution. 

The ö should be encoded as two bytes, i.e. 0x00f6 instead of 0xf6. Because it is not, `utf8::internal::sequence_length(it)` decides that it has to decode 4 bytes, which it cannot, because it only has one byte to work with.

**Comment by Manuel-K** [comment]
https://wiki.libsdl.org/SDL_TextInputEvent specifically describes text to be utf-8 encoded.

**Comment by Manuel-K** [comment]
00f6 is the unicode code position of ö. The hex representation is 0xc3b6. `utf8::internal::sequence_length(it)` returns 2 byte in this case, which is correct. So it seems that the function of utf8cpp that gets used to validate the input expects the hex representation, but it gets codepoints.

According to https://wiki.libsdl.org/MigrationGuide SDL should return utf-8 codepoints.

**Comment by Manuel-K** [comment]
@selaux 

Can you check what SDL_TextInputEvent.text you get on your system if you enter an ö?

**Comment by selaux** [comment]
I am getting, so two chars for a single ö, which seems valid utf-8 (didn't check the values though)

```
TextEv = {const SDL_TextInputEvent * | 0x7fffffffe110} 0x7fffffffe110
 type = {Uint32} 771
 timestamp = {Uint32} 5912
 windowID = {Uint32} 2
 text = {char [32]} 
  [0] = {char} -61 '\303'
  [1] = {char} -74 '\266'
  [2] = {char} 0 '\000'
  [3] = {char} 0 '\000'
  [4] = {char} 48 '0'
  [5] = {char} 1 '\001'
  [6] = {char} 0 '\000'
  [7] = {char} 65 'A'
  [8] = {char} 0 '\000'
  [9] = {char} 0 '\000'
  [10] = {char} 0 '\000'
  [11] = {char} 0 '\000'
  [12] = {char} 0 '\000'
  [13] = {char} -32 '\340'
  [14] = {char} 105 'i'
  [15] = {char} 62 '>'
  [16] = {char} 28 '\034'
  [17] = {char} 34 '"'
  [18] = {char} -115 '\215'
  [19] = {char} 105 'i'
  [20] = {char} -80 '\260'
  [21] = {char} -5 '\373'
  [22] = {char} -1 '\377'
  [23] = {char} 64 '@'
  [24] = {char} 0 '\000'
  [25] = {char} 0 '\000'
  [26] = {char} 0 '\000'
  [27] = {char} 0 '\000'
  [28] = {char} 108 'l'
  [29] = {char} -4 '\374'
  [30] = {char} -1 '\377'
  [31] = {char} 64 '@'
```

**Comment by Manuel-K** [comment]
0xC3B6 is 'ö'. 

I don't understand why your input is correct and I get the wrong value from SDL…

**Comment by selaux** [comment]
Might be worth raising an issue with SDL2 for that, because I think that this is out of our control.

**Comment by Manuel-K** [comment]
Yes, but I'd like to have a minimal test program that shows the same issue before I do that.

**Comment by Manuel-K** [comment]
I've just realized that I forgot to tell you about my research into this problem. I've written a minimal example that shows the same problem on my laptop (even when I'm logged in as a new user). I've compared locale, X11 and vconsole keymap settings to those on my desktop which does **not** show the same problem and haven't been able to spot a difference. The output of `localectl` is the same and `export` only has some minor differences (mostly process IDs).

---

A few minutes ago I've remembered the usual solution to similar problems of c# programs and ran the test with `LC_ALL=C`… this fixes the problem. Why?
